### PR TITLE
Leverage extension Debugger for Java to start/debug boot apps.

### PR DESCRIPTION
**Pull request by Eskibear**
Previously it supports "Start" by spawning a Java process. Now depending on "Debugger for Java" extension, it adds support for "Debug". Leveraging VS Code's "Start without debugging"  feature, the "Start" is also refactored in this PR.

See: #5 #7 

Brief introduction.
* The start/debug configuration (launch.json file) will be updated/created at first run. This also enables user to config vmargs, envs, cwd, etc. 
* It stops a boot app by sending "disconnect" request to the debug adapter in the corresponding debug session. (Exactly the same behavior with hitting stop button when debugging a generic java project).

Known issues:
* Debugger for Java doesn't support "Start without debugging" yet, causing it always run in debug mode. The feature is supposed to be shipped in 0.12.0 according to the [plan](https://github.com/Microsoft/vscode-java-debug/issues/357).  A workaround provided in this PR is to disable all BPs before "Start". 
* All the debug sessions share a single Debug Console in VS Code. This causes that all the logs for different applications are mixed together. See:  https://github.com/Microsoft/vscode/issues/46861 But there is a well-known workaround for it, one can always specifies console in `launch.json` to run apps in terminal instead of the internal debug console.
    ```
    "console": "integratedTerminal" or "externalTerminal"
    ```

### CODE REVIEW

**Comment by Eskibear** [comment]
Following is what it behaves with a private version of "Debugger for Java" installed. FYI. 

startndebug

File: src/Controller.ts [hunk 1]
```diff
@@ -20,47 +18,36 @@ export class Controller {
...
 
-    public async stopBootApp(app: BootApp): Promise<void> {
-        // TODO: How to send a shutdown signal to the app instead of killing the process directly?
-        if (app.process) {
-            app.process.kill("SIGTERM");
-            if (app.process.killed) {
-                app.process = undefined;
-            }
+    public async stopBootApp(app: BootApp, restart?: boolean): Promise<void> {
+        if (app.activeSession) {
```
Comment (kdvolder) [inline src/Controller.ts]: I suggest keeping this TODO. It is still valid/desirable.

**Comment by kdvolder** [comment]
I tried running it locally, when I tried to start an app I got a little problem.

This error popup appeared from the debugger about a build problem. (Annoying message, this is a known issue I have already reported: https://github.com/Microsoft/vscode-java-debug/issues/358).

However that isn't the problem. I clicked the 'Abort' button of the dialog. So, the app was not started. But boot dash showed it nevertheless as a green dot indicating a started app.

This something should be fixed before merging the PR? I can go either way on this, if you prefer we can treat it as a new bug ticket to be solved after the PR is merged.

File: src/Controller.ts (hunk 1 above)
Comment (Eskibear) [reply src/Controller.ts]: Sure, it was mistakenly deleted.

**Comment by Eskibear** [comment]
Thanks for the review. 
> I suggest putting back the TODO around proper shutdown.

Sure thing. 

> for the problem that if launch is aborted boot dash state shows the wrong running state for the app.

I just tested using your chatter repo, finding that `vscode.debug.startDebugging` returns "true" even though "Abort" is clicked. This looks strange. 
https://github.com/Microsoft/vscode-spring-boot-dashboard/blob/884f3cffbc1a9510f0e25bc7d8896c5126ac4fef/src/Controller.ts#L35-L42

Let me figure out how to fix it.

**Comment by Eskibear** [comment]
Someone else has the same problem using the API `vscode.debug.startDebugging`.
See: https://github.com/Microsoft/vscode/issues/54214

File: src/Controller.ts [hunk 2]
```diff
@@ -20,47 +18,36 @@ export class Controller {
...
-            javaProcess.on('exit', (code: number, signal: string) => {
-                this._setState(app, STATE_INACTIVE);
-            });
-            app.process = javaProcess;
+            const ok: boolean = await vscode.debug.startDebugging(
+                vscode.workspace.getWorkspaceFolder(vscode.Uri.parse(app.path)),
+                Object.assign({}, targetConfig, { noDebug: !debug })
+            );
+            if (ok) {
+                app.activeSession = vscode.debug.activeDebugSession;
```
Comment (Eskibear) [inline src/Controller.ts]: If there is a debug session running, when I try to start a new one but finally click "Abort",  `vscode.debug.startDebugging` still returns true, the previous running debug session is  still "vscode.debug.activeDebugSession", so it is wrongly set as current app's  activeSession.

**Comment by Eskibear** [comment]
Brief introduction for the latest commits.
Now a map<app.localtion, start/debugSession> is maintained in the BootAppManager.
* When starting an app, it constructs a launch configuration for the app, set the name of the session(not initialized yet) in the app, calls `startDebugging` API.
* When app is started, the session object is initialized, it set state for the corresponding app.
* When stopping an app, it get the session object by the session name stored in the app, then send `disconnect` request to the debug adapter through the session.
* When the app is stopped, the session object is also carried in the event, then we find the corresponding app, update the state.

PS:
Clicking "Abort" causes to open the launch.json file (create if not exists), that's VS Code's behavior when resolveDebugConfiguration  returns `undefined`. The message dialog for "Abort" is implemented in Java Debugger, I think the user experience of this one can be improved by adopting better logic in Java Debugger.

**Comment by Eskibear** [comment]
@kdvolder Could you help to take a look and test the fix?

**Comment by kdvolder** [comment]
> @kdvolder Could you help to take a look and test the fix?

Yes, of course, will try to get to it today.

**Pull request merged**